src/peergos/server/tests/simulation/Simulator.java

```diff
@@ -486,44 +558,79 @@ private boolean verifySharingPermissions(String user, Path path) {
                 LOG.info("User " + peergosFileSystem.user() + " path " + path + " has peergos-fs " + permission.name() + "ers " +
                         shareesInPeergos + " and local-fs " + permission.name() + "ers " + shareesInLocal);
                 isVerified = false;
+                break;
             }
 
             //check they can actually read
             for (String sharee : shareesInPeergos) {
-
                 byte[] read = null;
-                switch (permission) {
-                    case READ:
-                        try {
-                            //can read?
-                            PeergosFileSystemImpl fs = fileSystems.getTestFileSystem(sharee);
-                            read = fs.read(path);
-                        } catch (Exception ex) {
-                            LOG.log(Level.SEVERE, "User "+ sharee +" could not read shared-path "+ path +"!", ex);
-                            isVerified = false;
-                        }
-                        break;
-                    case WRITE:
-                        PeergosFileSystemImpl fs = fileSystems.getTestFileSystem(sharee);
-                        try {
-                            //can read?
-                            read = fs.read(path);
-                        } catch (Exception ex) {
-                            LOG.log(Level.SEVERE, "User "+ sharee +" could not read shared-path "+ path +"!", ex);
-                            isVerified = false;
-                        }
-                        if (isVerified) {
+                PeergosFileSystemImpl fs = fileSystems.getTestFileSystem(sharee);
+                if (isDir) {
```

Comment:
I  think forcing reading and writing  for  directories and files into two ops makes this quite complicated.

Suggestion: simpligy by removing the `READ_SHARED_DIR` and `WRITE_SHARED_DIR` ops and add new `MKDIR_SHARED` and  `LS_SHARED` operations that will create or list the contents  of  a (random) shared directory,  or raise an error if  path is a file.

Reply:
I'm reluctant to change this again. It does the job and i think it is clear. Can be enhanced in a subsequent PR.

Reply:
The fact that it is doing a mkdir and ls is just an implementation detail. Semantically it is testing whether we can read the directory, or write to the directory. And indeed we could change the implementation to for example delete the dir made. Instead of ls we could also check we have a read key in the Peergos case.